frame/contracts/src/wasm/runtime.rs

```diff
@@ -401,19 +445,24 @@ fn write_sandbox_output<E: Ext>(
 		Err(store_err(ctx, Error::<E::T>::OutputBufferTooSmall))?
 	}
 
-	charge_gas(
-		ctx.gas_meter,
-		ctx.schedule,
-		&mut ctx.trap_reason,
-		RuntimeToken::WriteMemory(buf_len.saturating_add(4)),
-	)?;
+	if let Some(token) = create_token(buf_len) {
+		charge_gas(ctx, token)?;
+	}
 
-	ctx.memory.set(out_ptr, buf)?;
-	ctx.memory.set(out_len_ptr, &buf_len.encode())?;
+	ctx.memory.set(out_ptr, buf).and_then(|_| {
+		ctx.memory.set(out_len_ptr, &buf_len.encode())
+	})
+	.map_err(|_| store_err(ctx, Error::<E::T>::OutOfBounds))?;
 
 	Ok(())
 }
 
+/// Can be supplied to `write_sandbox_output` to indicate that the gas meter is not be
+/// be charged for the copied data.
+fn no_charge(_: u32) -> Option<RuntimeToken> {
```

Comment:
Is this function **only** needed for the cases to avoid charging of additional weight for copying in cases where its already covered by the token?

Reply:
Yes. That is always the case when the size of the copy is static and not dependent on the input.

Reply:
I was wondering if "cost_already_accounted" or something like that, but I am not sure

Reply:
A renamed it to `already_charged` and updated its docs to more clearly describe its use case.